o3de/o3de
# Added tests for Asset Processor metrics database entries (#10518)

## What does this PR do?

This PR adds tests around changes in #10518 : Store and retrieve asset process job (`ProcessJob`) duration stats to/from Asset Database. Previously, only `AssetProcessor::AssetDatabaseConnection::ReplaceStat` is tested.

### AssetDatabaseTest
Ensure `AssetProcessor::AssetDatabaseConnection` APIs relevant to Stats table, namely `QueryStatsTable`, `GetStatByStatName`, `GetStatLikeStatName`, works correctly. There are helper functions in the test that helps creating stat entries (`GetFirstStatEntry`, `StepStatEntry`) and insert entries into the database (`InsertStatsTestData`). 

### JobModelTest
Added tests for `PopulateJobsFromDatabase()`.
- Ensure the job is populated from the database Jobs table (this piece of code existed before #10518).
- Ensure the stats are populated from the database Stats table.
- For stats that match jobs, ensure the job entry in the `JobsModel` is updated with the matched stat values.
- Ensure the stat that is ill-formatted is omitted, and `AZ_Warning` is emitted.

### StatsCapture
Added tests for `AssetProcessor::StatsCapture::EndCaptureStat(statName, persistToDb = false)`, which returns the just-made stat result and may store result to database.
- Ensure if the `statName` was seen before in a `AssetProcessor::StatsCapture::BeginCaptureStat` call, the just-made result will be returned (`AZStd::optional<>` has value). If the `statName` is not seen before, returned object has no value.
- Ensure if `persistToDb` is omitted or is false, the result is not written to database; if `persistToDb` is true, the result is written to database.

## How was this PR tested?

These are the tests that test the metrics persistence code.
I ran these tests locally and passed them.
